### ISSUE TITLE
rollup_candlestick() results

### ISSUE THREAD

**Issue by torwig** [VictoriaMetrics/VictoriaMetrics#309]
Version: 1.32.7

I have been testing the rollup_candlestick() function to get OHLC values.
I've written some data to VictoriaMetrics using the Influx line protocol (raw CURL requests attached). 
Measurement name: **asset**.
Random price values every **4min**.

1580802480 => price=9111.5 
1580802720 => price=9341.5 
1580802960 => price=9436.5 
**1580803200 => price=9384.5** (02/04/2020 @ 8:00am)
1580803440 => price=9300.4 
1580803680 => price=9287.6 
1580803920 => price=9316.2 
1580804160 => price=9423.5 
1580804400 => price=9574.1 
1580804640 => price=9600.9 
1580804880 => price=9579.4 
1580805120 => price=9624.2 
1580805360 => price=9740.5 
1580805600 => price=9642.1 
1580805840 => price=9500.1 
1580806080 => price=9450.2
1580806320 => price=9400.3 
1580806560 => price=9350.4 
**1580806800 => price=9300.5** (02/04/2020 @ 9:00am)
1580807040 => price=9250.6 
1580807280 => price=9150.8 
1580807520 => price=9100.9 
1580807760 => price=9240.5 
1580808000 => price=9564.7 
1580808240 => price=9532.7 
1580808480 => price=9421.7 
1580808720 => price=9325.7 
1580808960 => price=9514.7 
1580809200 => price=9647.7 
1580809440 => price=9874.7 
1580809680 => price=9742.7 
1580809920 => price=9532.7 
1580810160 => price=9412.7 
**1580810400 => price=9117.7** (02/04/2020 @ 10:00am)

I've got OHLC values over the interval from 8:00AM to 10:00AM with resolution of 1hour using the request:

`http://127.0.0.1:8428/api/v1/query_range?query=rollup_candlestick(asset_price)&start=1580803200&end=1580810400&step=3600`

And I've got the response:

{"status":"success","data":{"resultType":"matrix","result":
[
{"metric":{"rollup":"open","tag":"p"},
"values":[[1580803200,"9436.5"],[1580806800,"9300.4"],[1580810400,"9250.6"]]},
{"metric":{"rollup":"high","tag":"p"},
"values":[[1580803200,"9436.5"],[1580806800,"9740.5"],[1580810400,"9874.7"]]},
{"metric":{"rollup":"low","tag":"p"},
"values":[[1580803200,"9384.5"],[1580806800,"9287.6"],[1580810400,"9100.9"]]},
{"metric":{"rollup":"close","tag":"p"},
"values":[[1580803200,"9384.5"],[1580806800,"9300.5"],[1580810400,"9117.7"]]}
]
}}

In other words, I've received 3 OHLC candles:
[9436.5  9436.5  9384.5  9384.5] for 1580803200
[9300.4  9740.5  9287.6  9300.5] for 1580806800
[9250.6  9874.7  9100.9  9117.7] for 1580810400

The expected result should be in my opinion 2 OHLC candles:
[9384.5  9740.5  9287.6  9350.4] for 1580803200
[9300.5  9874.7  9100.9  9412.7] for 1580806800

It seems that VictoriaMetrics has divided data in the following intervals:

1580802960 => price=9436.5 (because of _maxSilenceInterval_ constant)
**1580803200 => price=9384.5** (02/04/2020 @ 8:00am)

1580803440 => price=9300.4 
1580803680 => price=9287.6 
1580803920 => price=9316.2 
1580804160 => price=9423.5 
1580804400 => price=9574.1 
1580804640 => price=9600.9 
1580804880 => price=9579.4 
1580805120 => price=9624.2 
1580805360 => price=9740.5 
1580805600 => price=9642.1 
1580805840 => price=9500.1 
1580806080 => price=9450.2 
1580806320 => price=9400.3 
1580806560 => price=9350.4 
**1580806800 => price=9300.5**

1580807040 => price=9250.6 
1580807280 => price=9150.8 
1580807520 => price=9100.9 
1580807760 => price=9240.5 
1580808000 => price=9564.7 
1580808240 => price=9532.7 
1580808480 => price=9421.7 
1580808720 => price=9325.7 
1580808960 => price=9514.7 
1580809200 => price=9647.7 
1580809440 => price=9874.7 
1580809680 => price=9742.7 
1580809920 => price=9532.7 
1580810160 => price=9412.7 
**1580810400 => price=9117.7** (02/04/2020 @ 10:00am)

And all the functions (rollupFirst, rollupLast, rollupMax, rollupMin) operate on these intervals.

Is my vision of the rollup_candlestick() result is wrong?
How should I use the rollup_candlestick() function to get OHLC values on the specified interval (from _start_ to _end_ values)?

[test_requests.txt](https://github.com/VictoriaMetrics/VictoriaMetrics/files/4153811/test_requests.txt)

**Comment by valyala** [comment]
VictoriaMetrics query engine uses Prometheus data model, which performs calculations on data points from the past. For instance, the returned results for timestamp 1580806800 are calculated over data points in the time range `(1580806800-step ... 1580806800]`, while you expect that results for this timestamp should be calculated on the time range `[1580806800 ... 1580806800+step)`.

The `rollup_candlestick` function implementation has been slightly changed in the commit ada6a3da8d8f97d3f130a217ad6f27cb62578910  in order to produce more expected results. Now it calculates results on the following time range for each timestamp `ts`: `[ts-step ... ts)`, i.e. value for `ts` isn't included in the calculation, while value for `ts-step` is included in the calculation. This should make results less misleading. But you need adding `offset -step` to the query in order to convert results to the expected range `[ts ... ts+step)` for each returned timestamp `ts`. I.e. the following query should be used:
```
http://127.0.0.1:8428/api/v1/query_range?query=rollup_candlestick(asset_price)+offset+-3600s&start=1580803200&end=1580810400&step=3600
```
The commit ada6a3da8d8f97d3f130a217ad6f27cb62578910 will be included in the next VictoriaMetrics release.

**Comment by valyala** [comment]
I think that adding `offset -step`  to the `rollup_candlestick()` query is too confusing, so this offset should be applied automatically starting from  the commit 525101339e9b30b45041f46689630921384fbcac . This means that the next VictoriaMetrics release should return the expected OHLC result from `rollup_candlestick()` function.

**Comment by valyala** [comment]
Fixes mentioned above are available in [v1.33.0](https://github.com/VictoriaMetrics/VictoriaMetrics/releases). @torwig , could you verify whether VictoriaMetrics [v1.33.0](https://github.com/VictoriaMetrics/VictoriaMetrics/releases) returns expected results for `rollup_candlestick` function?

**Comment by torwig** [comment]
@valyala , 
I've tried to run the following request using the v.1.33.0:

`http://127.0.0.1:8428/api/v1/query_range?query=rollup_candlestick(asset_price)&start=1580806800&end=1580810400&step=3600`

The idea was to get intervals [ 1580806800 - 3600 ... 1580806800 ) and 
[ 1580806800 ... 1580810400 ).
And here is the response:
{"status":"success","data":{"resultType":"matrix","result":[
{"metric":{"rollup":"low","tag":"p"},"values":[[1580806800,"9412.7"]]},
{"metric":{"rollup":"high","tag":"p"},"values":[[1580806800,"9412.7"]]},
{"metric":{"rollup":"open","tag":"p"},"values":[[1580806800,"9412.7"]]},
{"metric":{"rollup":"close","tag":"p"},"values":[[1580806800,"9412.7"]]}]
}}
which is not what I've expected.

I've also output the values on which rollupOpen/rollupClose/rollupHigh/rollupLow were operating:
rfa.currTimestamp= 1580810400000
rollupOpen() values [9412.7]
rfa.currTimestamp= 1580814000000
rollupOpen() values []
rfa.currTimestamp= 1580817600000
rollupOpen() values []
rfa.currTimestamp= 1580810400000
rollupClose() values [9412.7]
rfa.currTimestamp= 1580814000000
rollupClose() values []
rfa.currTimestamp= 1580817600000
rollupClose() values []
rfa.currTimestamp= 1580810400000
rollupLow() values [9412.7]
rfa.currTimestamp= 1580814000000
rollupLow() values []
rfa.currTimestamp= 1580817600000
rollupLow() values []
rfa.currTimestamp= 1580810400000
rollupHigh() values [9412.7]
rfa.currTimestamp= 1580814000000
rollupHigh() values []
rfa.currTimestamp= 1580817600000
rollupHigh() values []

I've inserted this code at line 442 of eval.go:

`fmt.Printf("ecNew.Start=%d, ecNew.End=%d, step=%d, offset=%d\n", ecNew.Start, ecNew.End, ecNew.Step, offset)`

And here is the output:

`ecNew.Start=1580810400000, ecNew.End=1580817600000, step=3600000, offset=-3600000`

**Comment by valyala** [comment]
@torwig , could you try building VictoriaMetrics from the latest commit and verifying whether the issue has been fixed? Note that now VictoriaMetrics should return the expected OHLC results for timestamp `ts` on the range `[ts ... ts+step)`, so yo don't need to adjust the resulting timestamps with `offset -step` or something similar. See [how to build VictoriaMetrics from sources](https://github.com/VictoriaMetrics/VictoriaMetrics/blob/master/README.md#how-to-build-from-sources).

**Comment by torwig** [comment]
@valyala ,
I've built VM from the source and made 2 requests.

The 1st one is:

`http://127.0.0.1:8428/api/v1/query_range?query=rollup_candlestick(asset_price)&start=1580806800&end=1580810400&step=3600`

And here is the response:

{"status":"success","data":{"resultType":"matrix","result":[
{"metric":{"rollup":"low","tag":"p"},"values":[[1580806800,"9100.9"]]},
{"metric":{"rollup":"high","tag":"p"},"values":[[1580806800,"9874.7"]]},
{"metric":{"rollup":"open","tag":"p"},"values":[[1580806800,"9300.5"]]},
{"metric":{"rollup":"close","tag":"p"},"values":[[1580806800,"9412.7"]]}
]
}}

The 2nd one is:

`http://127.0.0.1:8428/api/v1/query_range?query=rollup_candlestick(asset_price)&start=1580803200&end=1580810400&step=3600`

And here is the response:

{"status":"success","data":{"resultType":"matrix","result":[
{"metric":{"rollup":"low","tag":"p"},"values":[[1580803200,"9287.6"],[1580806800,"9100.9"]]},
{"metric":{"rollup":"high","tag":"p"},"values":[[1580803200,"9740.5"],[1580806800,"9874.7"]]},
{"metric":{"rollup":"open","tag":"p"},"values":[[1580803200,"9384.5"],[1580806800,"9300.5"]]},
{"metric":{"rollup":"close","tag":"p"},"values":[[1580803200,"9350.4"],[1580806800,"9412.7"]]}
]}}

It seems everything works OK now. 
I'll also double-check tomorrow on real data.
Thanks!

**Comment by torwig** [comment]
@valyala ,
I've tested on real data and your fix works fine.

**Comment by valyala** [comment]
Great! Then the fix will be included in the next release.

**Comment by valyala** [comment]
The bugfix is available starting from [v1.33.1](https://github.com/VictoriaMetrics/VictoriaMetrics/releases). Closing the issue as fixed.

**Comment by torwig** [comment]
@valyala 
I've found a few issues and described them here:
https://github.com/VictoriaMetrics/VictoriaMetrics/pull/328

**Comment by torwig** [comment]
@valyala 
Thank you.
And could you please somehow check the intervals on which rollup_candlestick() operates because it looks like the function operates on `(ts, ts + step)` rather than on `[ts, ts + step)`?

In my case I have the following values:
price=9782.000000 **1582532760**000000000
price=9781.900000 1582532768000000000
price=9781.900000 1582532773000000000
price=9782.000000 1582532777000000000
price=9782.000000 1582532777001000000
price=9782.000000 1582532777002000000
price=9784.900000 1582532806000000000
price=9784.900000 1582532806001000000
And here is the result:
...{"metric":{"rollup":"open","topic":"trade"},"values":[[**1582532760**,"9781.9"].
It occurs only when **ts** divisible by **step**.

**Comment by valyala** [comment]
@torwig , this should be fixed in ae51300973d4aab42fe5d180e4323eaf9ba4b25e .

**Comment by torwig** [comment]
@valyala 
Yes, it has been fixed.

**Comment by valyala** [comment]
The fix has been included in [v1.34.0](https://github.com/VictoriaMetrics/VictoriaMetrics/releases)
